File: previous-iteration/appmap/validate.py
"""Validation — the linter's teeth, but warn-only. Never blocks; exit 0.

Three checks:
  * frontmatter shape vs surface.schema.json (hand-rolled subset — no jsonschema dep),
  * code_anchor resolution (file exists; symbol appears in it),
  * screenshot path resolution (state screenshots resolve to a file),
plus dangling-link reporting (from links.build_links).
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from .config import Config
from .links import build_links
from .model import Surface

WARN = "warn"
ERROR = "error"  # still non-blocking; severity is for the human's attention only


@dataclass
class Finding:
    level: str
    surface: str
    message: str


# ── tiny JSON-schema-subset validator (required/type/enum/pattern) ───────────

_TYPE_CHECKS = {
    "string": lambda v: isinstance(v, str),
    "boolean": lambda v: isinstance(v, bool),
    "array": lambda v: isinstance(v, list),
    "object": lambda v: isinstance(v, dict),
    "number": lambda v: isinstance(v, (int, float)) and not isinstance(v, bool),
}


def _check_node(value: Any, schema: dict[str, Any], loc: str, out: list[str]) -> None:
    t = schema.get("type")
    if t and t in _TYPE_CHECKS and value is not None and not _TYPE_CHECKS[t](value):
        out.append(f"{loc}: expected {t}, got {type(value).__name__}")
        return

    if t == "object" and isinstance(value, dict):
        for req in schema.get("required", []):
            if req not in value:
                out.append(f"{loc}: missing required '{req}'")
        props = schema.get("properties", {})
        for k, sub in props.items():
            if k in value and value[k] is not None:
                _check_node(value[k], sub, f"{loc}.{k}" if loc else k, out)

    if t == "array" and isinstance(value, list):
        item_schema = schema.get("items")
        if item_schema:
            for i, item in enumerate(value):
                _check_node(item, item_schema, f"{loc}[{i}]", out)

    if isinstance(value, str):
        enum = schema.get("enum")
        if enum and value not in enum:
            out.append(f"{loc}: '{value}' not one of {enum}")
        pat = schema.get("pattern")
        if pat and not re.match(pat, value):
            out.append(f"{loc}: '{value}' does not match /{pat}/")


def _schema_findings(surface: Surface, schema: dict[str, Any]) -> list[Finding]:
    msgs: list[str] = []
    _check_node(surface.data, schema, "", msgs)
    return [Finding(WARN, surface.id, m) for m in msgs]


# ── code anchor + screenshot resolution ──────────────────────────────────────

def _anchor_findings(surface: Surface, cfg: Config) -> list[Finding]:
    out: list[Finding] = []
    anchor = surface.code_anchor
    if not anchor:
        return out
    file_rel = anchor.get("file")
    if not file_rel:
        return out
    target = (cfg.project_root / file_rel)
    if not target.is_file():
        out.append(Finding(ERROR, surface.id, f"code_anchor.file missing: {file_rel}"))
        return out
    symbol = anchor.get("symbol")
    if symbol:
        try:
            text = target.read_text(encoding="utf-8", errors="ignore")
            if not re.search(rf"\b{re.escape(symbol)}\b", text):
                out.append(
                    Finding(WARN, surface.id, f"code_anchor.symbol '{symbol}' not found in {file_rel}")
                )
        except Exception as e:  # pragma: no cover - defensive
            out.append(Finding(WARN, surface.id, f"could not read {file_rel}: {e}"))
    return out


def _screenshot_findings(surface: Surface, cfg: Config) -> list[Finding]:
    out: list[Finding] = []
    surface_dir = surface.path.parent
    search_dirs = [surface_dir] + [cfg.project_root / d for d in cfg.screenshot_dirs]
    for st in surface.states:
        shot = st.get("screenshot")
        if not shot:
            continue
        if any((d / shot).is_file() or Path(shot).is_absolute() and Path(shot).is_file()
               for d in search_dirs):
            continue
        out.append(
            Finding(WARN, surface.id,
                    f"state '{st.get('name', '?')}' screenshot unresolved: {shot}")
        )
    return out


# ── entrypoint ───────────────────────────────────────────────────────────────

def validate(surfaces: list[Surface], cfg: Config) -> list[Finding]:
    findings: list[Finding] = []

    schema: dict[str, Any] = {}
    if cfg.schema_path.is_file():
        try:
            schema = json.loads(cfg.schema_path.read_text(encoding="utf-8"))
        except Exception as e:
            findings.append(Finding(WARN, "-", f"could not load schema: {e}"))

    for s in surfaces:
        if schema:
            findings += _schema_findings(s, schema)
        findings += _anchor_findings(s, cfg)
        findings += _screenshot_findings(s, cfg)

    links = build_links(surfaces)
    for d in links.dangling:
        findings.append(
            Finding(ERROR, d.from_id, f"dangling {d.kind} -> '{d.to}' (no such surface) [{d.ref}]")
        )

    return findings

File: cli/Sources/AppMapKit/Validate.swift
// Validation — the linter's teeth, but warn-only. Never blocks; the CLI exits
// 0 regardless of findings. Severity ranks attention only.
//
// Checks:
//   * frontmatter shape vs surface.schema.json (hand-rolled subset:
//     required/type/enum/pattern — no JSON-schema dependency),
//   * code_anchor resolution (file exists; symbol appears, word-bounded),
//   * screenshot path resolution (state screenshots resolve to a file),
//   * dangling links (edge/contains/entry_point targets with no record).

import Foundation

public enum FindingLevel: String {
    case warn
    case error  // still non-blocking; severity is for the human's attention only
}

public struct Finding {
    public let level: FindingLevel
    public let surface: String
    public let message: String

    public init(_ level: FindingLevel, _ surface: String, _ message: String) {
        self.level = level
        self.surface = surface
        self.message = message
    }
}

// ── tiny JSON-schema-subset validator (required/type/enum/pattern) ──────────

func typeMatches(_ value: Any, _ type: String) -> Bool {
    switch type {
    case "string": return value is String
    case "boolean": return value is Bool
    case "array": return value is [Any]
    case "object": return asStringDict(value) != nil
    case "number": return (value is Int || value is Double) && !(value is Bool)
    default: return true
    }
}

func isNull(_ value: Any) -> Bool { value is NSNull }

func checkNode(_ value: Any, schema: [String: Any], loc: String, out: inout [String]) {
    let type = schema["type"] as? String
    if let type, !isNull(value), !typeMatches(value, type) {
        out.append("\(loc): expected \(type), got \(yamlTypeName(value))")
        return
    }

    if type == "object", let dict = asStringDict(value) {
        for req in (schema["required"] as? [Any])?.compactMap({ $0 as? String }) ?? [] {
            if dict[req] == nil {
                out.append("\(loc): missing required '\(req)'")
            }
        }
        let props = asStringDict(schema["properties"]) ?? [:]
        for (key, sub) in props.sorted(by: { $0.key < $1.key }) {
            guard let subSchema = asStringDict(sub),
                  let subValue = dict[key], !isNull(subValue) else { continue }
            checkNode(subValue, schema: subSchema, loc: loc.isEmpty ? key : "\(loc).\(key)", out: &out)
        }
    }

    if type == "array", let items = value as? [Any],
       let itemSchema = asStringDict(schema["items"]) {
        for (i, item) in items.enumerated() {
            checkNode(item, schema: itemSchema, loc: "\(loc)[\(i)]", out: &out)
        }
    }

    if let str = value as? String {
        if let anyEnum = schema["enum"] as? [Any] {
            let allowed = anyEnum.compactMap { $0 as? String }
            if !allowed.contains(str) {
                out.append("\(loc): '\(str)' not one of \(allowed)")
            }
        }
        if let pattern = schema["pattern"] as? String, !matchesFromStart(str, pattern: pattern) {
            out.append("\(loc): '\(str)' does not match /\(pattern)/")
        }
    }
}

func yamlTypeName(_ value: Any) -> String {
    if value is Bool { return "boolean" }
    if value is Int || value is Double { return "number" }
    if value is String { return "string" }
    if value is [Any] { return "array" }
    if asStringDict(value) != nil { return "object" }
    return String(describing: Swift.type(of: value))
}

func matchesFromStart(_ value: String, pattern: String) -> Bool {
    guard let re = try? NSRegularExpression(pattern: pattern) else { return true }
    let range = NSRange(value.startIndex..., in: value)
    guard let m = re.firstMatch(in: value, range: range) else { return false }
    return m.range.location == 0
}

func schemaFindings(_ surface: SurfaceRecord, schema: [String: Any]) -> [Finding] {
    var msgs: [String] = []
    checkNode(surface.data, schema: schema, loc: "", out: &msgs)
    return msgs.map { Finding(.warn, surface.id, $0) }
}

// ── code anchor + screenshot resolution ─────────────────────────────────────

func anchorFindings(_ surface: SurfaceRecord, cfg: Config) -> [Finding] {
    let anchor = surface.codeAnchor
    guard let fileRel = anchor["file"] as? String, !fileRel.isEmpty else { return [] }
    let target = cfg.projectRoot.appendingPathComponent(fileRel)
    guard FileManager.default.fileExists(atPath: target.path) else {
        return [Finding(.error, surface.id, "code_anchor.file missing: \(fileRel)")]
    }
    guard let symbol = anchor["symbol"] as? String, !symbol.isEmpty else { return [] }
    guard let text = try? String(contentsOf: target, encoding: .utf8) else {
        return [Finding(.warn, surface.id, "could not read \(fileRel)")]
    }
    let escaped = NSRegularExpression.escapedPattern(for: symbol)
    let re = try? NSRegularExpression(pattern: "\\b\(escaped)\\b")
    let found = re?.firstMatch(in: text, range: NSRange(text.startIndex..., in: text)) != nil
    if !found {
        return [Finding(.warn, surface.id, "code_anchor.symbol '\(symbol)' not found in \(fileRel)")]
    }
    return []
}

/// A watches entry pointing at a vanished file is a hole in the drift net —
/// same class of dead reference as a missing anchor file.
func watchesFindings(_ surface: SurfaceRecord, cfg: Config) -> [Finding] {
    surface.watches
        .filter { !FileManager.default.fileExists(atPath: cfg.projectRoot.appendingPathComponent($0).path) }
        .map { Finding(.error, surface.id, "watches file missing: \($0)") }
}

func screenshotFindings(_ surface: SurfaceRecord, cfg: Config) -> [Finding] {
    var out: [Finding] = []
    let fm = FileManager.default
    let surfaceDir = surface.path.deletingLastPathComponent()
    let searchDirs = [surfaceDir] + cfg.screenshotDirs.map {
        cfg.projectRoot.appendingPathComponent($0)
    }
    for state in surface.states {
        guard let shot = state["screenshot"] as? String, !shot.isEmpty else { continue }
        let resolved = searchDirs.contains {
            fm.fileExists(atPath: $0.appendingPathComponent(shot).path)
        } || (shot.hasPrefix("/") && fm.fileExists(atPath: shot))
        if !resolved {
            let name = state["name"] as? String ?? "?"
            out.append(Finding(.warn, surface.id, "state '\(name)' screenshot unresolved: \(shot)"))
        }
    }
    return out
}

// ── entrypoint ───────────────────────────────────────────────────────────────

public func loadSchema(_ cfg: Config) -> ([String: Any]?, Finding?) {
    guard FileManager.default.fileExists(atPath: cfg.schemaPath.path) else { return (nil, nil) }
    do {
        let data = try Data(contentsOf: cfg.schemaPath)
        let obj = try JSONSerialization.jsonObject(with: data)
        guard let dict = obj as? [String: Any] else {
            return (nil, Finding(.warn, "-", "could not load schema: not a JSON object"))
        }
        return (dict, nil)
    } catch {
        return (nil, Finding(.warn, "-", "could not load schema: \(error.localizedDescription)"))
    }
}

public func validate(_ surfaces: [SurfaceRecord], cfg: Config) -> [Finding] {
    var findings: [Finding] = []

    let (schema, schemaLoadFinding) = loadSchema(cfg)
    if let schemaLoadFinding { findings.append(schemaLoadFinding) }

    for s in surfaces {
        if let schema { findings += schemaFindings(s, schema: schema) }
        findings += anchorFindings(s, cfg: cfg)
        findings += watchesFindings(s, cfg: cfg)
        findings += screenshotFindings(s, cfg: cfg)
    }

    let links = buildLinks(surfaces)
    for d in links.dangling {
        findings.append(Finding(
            .error, d.fromID,
            "dangling \(d.kind) -> '\(d.to)' (no such surface) [\(d.ref)]"
        ))
    }
    return findings
}
